File: lintcode/639.word-abbreviation.py
class Solution:
    """
    @param dict: an array of n distinct non-empty strings
    @return: an array of minimal possible abbreviations for every word
    """

    def wordsAbbreviation(self, words):
        def abbrev(word, end=0):
            return word[0:end + 1] + str(len(word) - 2 - end) + word[-1]

        duplicates = {}
        for word in words:
            aword = abbrev(word)
            print(f'{word} -> {aword}')

            if aword not in duplicates:
                duplicates[aword] = []

            duplicates[aword].append(word)
        import pprint
        pprint.pprint(duplicates)

        final_map = {}

        class Trie:
            def __init__(self, val):
                self.children = {}
                self.count = 0

        for (k, v) in duplicates.items():
            if len(v) == 1:
                final_map[v[0]] = k
                continue

            dummy = Trie('')
            for aw in v:
                prev = dummy

                for (_, ch) in enumerate(aw):
                    if ch not in prev.children:
                        prev.children[ch] = Trie(ch)
                    prev.children[ch].count += 1
                    prev = prev.children[ch]

            for aw in v:
                prev = dummy

                for (idx, ch) in enumerate(aw):
                    curr = prev.children[ch]
                    if curr.count == 1:
                        final_map[aw] = abbrev(aw, idx)
                        break
                    prev = curr

        assert len(final_map) == len(words)

        res = []
        for word in words:
            new_word = final_map[word]
            if len(new_word) >= len(word):
                res.append(word)
            else:
                res.append(new_word)
        print(res)
        return res


if __name__ == '__main__':
    Solution().wordsAbbreviation([
        "like", "god", "internal", "me", "internet", "interval", "intension",
        "face", "intrusion"
    ])

File: lintcode/s/639.word-abbreviation.cpp
#include <memory>
#include <string>
#include <vector>
using namespace std;

struct TrieNode {
  unordered_map<char, shared_ptr<TrieNode>> children;
  int count;

  TrieNode() : count(0) {}
};

class Solution {
  string abbrev(const string &str, int i) {
    string res = str.substr(0, i);
    int len = str.size() - i - 1;
    res += to_string(len);
    res += str.back();

    return res.size() < str.size() ? res : str;
  }

public:
  /**
   * @param dict: an array of n distinct non-empty strings
   * @return: an array of minimal possible abbreviations for every word
   */
  vector<string> wordsAbbreviation(vector<string> &dict) {
    unordered_map<string, vector<string>> groups;

    for (auto const &word : dict) {
      string aWord = abbrev(word, 1);
      groups[aWord].push_back(word);
    }

    unordered_map<string, string> res;
    for (auto group : groups) {
      auto aWord = group.first;
      auto words = group.second;

      if (words.size() == 1) {
        res[words[0]] = aWord;
        continue;
      }

      shared_ptr<TrieNode> root = make_shared<TrieNode>();
      for (string word : words) {
        auto curr = root;
        for (char ch : word) {
          if (!curr->children.count(ch)) {
            curr->children[ch] = make_shared<TrieNode>();
          }
          curr = curr->children[ch];
          curr->count += 1;
        }
      }

      for (string word : words) {
        auto curr = root;
        for (int i = 0; i < word.size(); ++i) {
          char ch = word[i];
          curr = curr->children[ch];
          if (curr->count == 1) {
            res[word] = abbrev(word, i + 1);
            break;
          }
        }
      }
    }
    vector<string> ret;
    for (auto const &str : dict) {
      ret.push_back(res[str]);
    }
    return ret;
  }
};
